chain-abci/src/app/rewards.rs

```diff
@@ -0,0 +1,193 @@
+use std::cmp::min;
+use std::f64::consts::E;
+
+use chain_core::init::{coin::Coin, params::NetworkParameters};
+use chain_core::state::account::StakedStateAddress;
+use chain_core::state::tendermint::TendermintValidatorAddress;
+use chain_core::tx::fee::Milli;
+
+use crate::app::{update_account, ChainNodeApp};
+use crate::enclave_bridge::EnclaveProxy;
+use crate::storage::tx::get_account;
+
+fn milli_to_f64(milli: Milli) -> f64 {
+    milli.as_millis() as f64 / 1000_f64
+}
+
+/// FIXME more stable arithmetic
+pub fn monetary_expansion(
+    tau: Milli,
+    total_staking: Coin,
+    minted: Coin,
+    params: &NetworkParameters,
+) -> Coin {
+    let cap = params.get_rewards_monetary_expansion_cap();
+    let r0 = milli_to_f64(params.get_rewards_monetary_expansion_r0());
+    let tau = milli_to_f64(tau);
+    let total_staking = u64::from(total_staking) as f64 / 1_0000_0000_f64;
+    let amount = total_staking * r0 * E.powf(-total_staking / tau);
+    min(
+        (cap - minted).unwrap_or_default(),
+        Coin::new(amount as u64).unwrap(),
+    )
+}
+
+type RewardsDistribution = Vec<(StakedStateAddress, Coin)>;
+impl<T: EnclaveProxy> ChainNodeApp<T> {
+    /// Distribute rewards pool
+    pub fn rewards_try_distribute(&mut self) -> Option<(RewardsDistribution, Coin)> {
+        let state = self.last_state.as_mut().unwrap();
+
+        if state.block_time < state.genesis_time
+            || state.block_time < state.rewards_pool.last_distribution_time
+        {
+            // FIXME use global overflow/underflow check.
+            panic!("invalid block time");
+        }
+
+        if state.block_time - state.rewards_pool.last_distribution_time
+            < state.network_params.get_rewards_distribution_period() as u64
+        {
+            return None;
+        }
+        self.rewards_pool_updated = true;
+
+        let mut total_staking = Coin::zero();
+        for (addr, _) in self.validator_voting_power.iter() {
+            let account = get_account(addr, &state.last_account_root_hash, &self.accounts)
+                .expect("io error or validator account not exists");
+            total_staking = (total_staking + account.bonded).expect("coin overflow");
+        }
+
+        let minted = monetary_expansion(
```

Comment:
one sanity check here to prevent infinite minting -- `min(state.rewards_pool.remaining, monetary_expansion(..))`

Reply:
All of `minted + remaining` rewards will get distributed in the current period.